Subject: [PATCH] new patch
---
Index: src/main/java/net/engineeringdigest/journalApp/Controller/PublicController.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/Controller/PublicController.java b/src/main/java/net/engineeringdigest/journalApp/Controller/PublicController.java
new file mode 100644
--- /dev/null	(date 1763054307042)
+++ b/src/main/java/net/engineeringdigest/journalApp/Controller/PublicController.java	(date 1763054307042)
@@ -0,0 +1,24 @@
+package net.engineeringdigest.journalApp.Controller;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.services.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/public")
+public class PublicController {
+
+    @Autowired
+    private UserService userService;
+
+
+    @PostMapping("/create-user")
+    public void create(@RequestBody User user) {
+
+        userService.saveNewUser(user);
+    }
+}
Index: src/main/java/net/engineeringdigest/journalApp/Controller/UserController.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/Controller/UserController.java b/src/main/java/net/engineeringdigest/journalApp/Controller/UserController.java
new file mode 100644
--- /dev/null	(date 1763054307076)
+++ b/src/main/java/net/engineeringdigest/journalApp/Controller/UserController.java	(date 1763054307076)
@@ -0,0 +1,57 @@
+package net.engineeringdigest.journalApp.Controller;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.repositry.UserEntryRepository;
+import net.engineeringdigest.journalApp.services.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/users")
+public class UserController {
+
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private UserEntryRepository userRepository;
+
+    @GetMapping
+    public List<User> getAllUser(){
+         return userService.getAll();
+    }
+
+
+    @PostMapping
+    public void create(@RequestBody User user) {
+
+        userService.saveNewUser(user);
+    }
+
+    @PutMapping
+    public ResponseEntity<?> updateUser(@RequestBody User user ){
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        String userName = authentication.getName();
+        User userInDb = userService.findByUserName(userName);
+        userInDb.setUserName(user.getUserName());
+        userInDb.setPassword(user.getPassword());
+        userService.saveNewUser(userInDb);
+        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+    }
+
+    @DeleteMapping
+   public ResponseEntity<?> deleteUserById(){
+       Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+       userRepository.deleteByUserName(authentication.getName());
+       return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+   }
+
+
+
+}
Index: src/main/java/net/engineeringdigest/journalApp/services/UserService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/services/UserService.java b/src/main/java/net/engineeringdigest/journalApp/services/UserService.java
new file mode 100644
--- /dev/null	(date 1763359506246)
+++ b/src/main/java/net/engineeringdigest/journalApp/services/UserService.java	(date 1763359506246)
@@ -0,0 +1,63 @@
+package net.engineeringdigest.journalApp.services;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.repositry.UserEntryRepository;
+import org.bson.types.ObjectId;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class UserService {
+
+    @Autowired
+    private UserEntryRepository userEntryRepository;
+
+    private static final PasswordEncoder passwordEncoder=new BCryptPasswordEncoder();
+
+    public boolean saveNewUser(User user){
+
+        try {
+            user.setPassword(passwordEncoder.encode(user.getPassword()));
+            user.setRoles(Arrays.asList("USER"));
+            userEntryRepository.save(user);
+            return true;
+        }catch (Exception e){
+            return false;
+        }
+
+    }
+
+    public void saveUser(User user){
+        userEntryRepository.save(user);
+    }
+
+    public List<User> getAll(){
+        return userEntryRepository.findAll();
+    }
+
+    public Optional<User> findById(ObjectId id){
+        return userEntryRepository.findById(id);
+    }
+
+
+    public void deleteById(ObjectId id){
+        userEntryRepository.deleteById(id);
+    }
+
+    public User findByUserName(String userName){
+       return userEntryRepository.findByUserName(userName);
+    }
+
+
+    public void saveAdmin(User user) {
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        user.setRoles(Arrays.asList("ADMIN"));
+        userEntryRepository.save(user);
+    }
+}
Index: src/main/java/net/engineeringdigest/journalApp/repositry/UserEntryRepository.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/repositry/UserEntryRepository.java b/src/main/java/net/engineeringdigest/journalApp/repositry/UserEntryRepository.java
new file mode 100644
--- /dev/null	(date 1762951123336)
+++ b/src/main/java/net/engineeringdigest/journalApp/repositry/UserEntryRepository.java	(date 1762951123336)
@@ -0,0 +1,14 @@
+package net.engineeringdigest.journalApp.repositry;
+
+import net.engineeringdigest.journalApp.entity.JournalEntry;
+import net.engineeringdigest.journalApp.entity.User;
+import org.bson.types.ObjectId;
+import org.springframework.data.mongodb.repository.MongoRepository;
+
+public interface UserEntryRepository extends MongoRepository<User, ObjectId> {
+
+    User findByUserName(String userName);
+
+    void deleteByUserName(String userName);
+
+}
Index: src/test/java/net/engineeringdigest/journalApp/service/UserServiceImplTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/net/engineeringdigest/journalApp/service/UserServiceImplTest.java b/src/test/java/net/engineeringdigest/journalApp/service/UserServiceImplTest.java
new file mode 100644
--- /dev/null	(date 1763378149708)
+++ b/src/test/java/net/engineeringdigest/journalApp/service/UserServiceImplTest.java	(date 1763378149708)
@@ -0,0 +1,41 @@
+package net.engineeringdigest.journalApp.service;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.repositry.UserEntryRepository;
+import net.engineeringdigest.journalApp.services.UserDetailsServiceImpl;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.ArrayList;
+
+import static org.mockito.Mockito.when;
+
+@SpringBootTest
+public class UserServiceImplTest {
+
+    @InjectMocks
+    private UserDetailsServiceImpl userDetailsService;
+
+    @MockBean
+    private UserEntryRepository userEntryRepository;
+
+    @BeforeEach
+    void setup(){
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    void loadUserByUserNameTest(){
+
+        when(userEntryRepository.findByUserName(ArgumentMatchers.anyString())).thenReturn(User.builder().userName("ram").password("intredv").roles(new ArrayList<>()).build());
+        UserDetails user = userDetailsService.loadUserByUsername("ram");
+        Assertions.assertNotNull(user);
+
+    }
+}
Index: src/main/java/net/engineeringdigest/journalApp/entity/User.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/entity/User.java b/src/main/java/net/engineeringdigest/journalApp/entity/User.java
new file mode 100644
--- /dev/null	(date 1763361738959)
+++ b/src/main/java/net/engineeringdigest/journalApp/entity/User.java	(date 1763361738959)
@@ -0,0 +1,30 @@
+package net.engineeringdigest.journalApp.entity;
+
+import com.sun.istack.internal.NotNull;
+import lombok.Builder;
+import lombok.Data;
+import org.bson.types.ObjectId;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.index.Indexed;
+import org.springframework.data.mongodb.core.mapping.DBRef;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Document
+@Data
+@Builder
+public class User {
+    @Id
+    private ObjectId id;
+    @Indexed(unique = true)
+    @NotNull
+    private String userName;
+    @NotNull
+    private String password;
+
+    @DBRef
+    private List<JournalEntry> journalEntries=new ArrayList<>();
+    private List<String> roles;
+}
Index: src/main/java/net/engineeringdigest/journalApp/Controller/AdminController.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/Controller/AdminController.java b/src/main/java/net/engineeringdigest/journalApp/Controller/AdminController.java
new file mode 100644
--- /dev/null	(date 1763197768947)
+++ b/src/main/java/net/engineeringdigest/journalApp/Controller/AdminController.java	(date 1763197768947)
@@ -0,0 +1,34 @@
+package net.engineeringdigest.journalApp.Controller;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.services.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/admin")
+public class AdminController {
+
+    @Autowired
+    public UserService userService;
+
+    @GetMapping("/all-users")
+    public ResponseEntity<?> getAllUser(){
+        List<User> all = userService.getAll();
+        if(all !=null && !all.isEmpty()){
+            return new ResponseEntity<>(all, HttpStatus.OK);
+        }
+        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+    }
+
+    @PostMapping("/create-admin-user")
+    public void createUser(@RequestBody  User user) {
+        userService.saveAdmin(user);
+
+    }
+
+}
Index: src/main/resources/application.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>ISO-8859-1
===================================================================
diff --git a/src/main/resources/application.properties b/src/main/resources/application.properties
new file mode 100644
--- /dev/null	(date 1763222565124)
+++ b/src/main/resources/application.properties	(date 1763222565124)
@@ -0,0 +1,8 @@
+spring.data.mongodb.uri= mongodb+srv://cjagtap7588_db_user:O8mqSl3raWWVOUiR@cluster1.7aaxb3q.mongodb.net/?appName=Cluster1
+spring.data.mongodb.database=journalDB
+#spring.data.mongodb.username=dba
+#spring.data.mongodb.password=ds
+
+spring.data.mongodb.auto-index-creation=true
+#spring.data.mongodb.ssl.enabled=true
+server.port=8081
Index: src/main/java/net/engineeringdigest/journalApp/services/UserDetailsServiceImpl.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/services/UserDetailsServiceImpl.java b/src/main/java/net/engineeringdigest/journalApp/services/UserDetailsServiceImpl.java
new file mode 100644
--- /dev/null	(date 1762947724561)
+++ b/src/main/java/net/engineeringdigest/journalApp/services/UserDetailsServiceImpl.java	(date 1762947724561)
@@ -0,0 +1,30 @@
+package net.engineeringdigest.journalApp.services;
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.repositry.UserEntryRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserDetailsServiceImpl implements UserDetailsService {
+    @Autowired
+    private UserEntryRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+
+        User user = userRepository.findByUserName(username);
+        if(user != null){
+            return org.springframework.security.core.userdetails.User.builder()
+                    .username(user.getUserName())
+                    .password(user.getPassword())
+                    .roles(user.getRoles().toArray(new String[0]))
+                    .build();
+
+        }
+        throw new UsernameNotFoundException("user not found with username" + username);
+    }
+}
Index: src/test/java/net/engineeringdigest/journalApp/service/UserServiceTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/net/engineeringdigest/journalApp/service/UserServiceTest.java b/src/test/java/net/engineeringdigest/journalApp/service/UserServiceTest.java
new file mode 100644
--- /dev/null	(date 1763363062052)
+++ b/src/test/java/net/engineeringdigest/journalApp/service/UserServiceTest.java	(date 1763363062052)
@@ -0,0 +1,49 @@
+package net.engineeringdigest.journalApp.service;
+
+
+import net.engineeringdigest.journalApp.entity.User;
+import net.engineeringdigest.journalApp.repositry.UserEntryRepository;
+
+import net.engineeringdigest.journalApp.services.UserService;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
+public class UserServiceTest {
+
+    @Autowired
+    private UserEntryRepository userRepository;
+
+    @Autowired
+    private UserService userService;
+
+//    @ParameterizedTest
+//    @ArgumentsSource(UserArgumentProvider.class)
+//    public void testSaveNewUser(User user){
+//
+//
+//        assertTrue(userService.saveNewUser(user));
+//
+//    }
+
+
+
+    @Disabled
+    @ParameterizedTest
+    @CsvSource({
+            "2,3,5",
+            "2,4,6",
+            "1,2,1"
+    })
+    public void test(int a, int b, int expected){
+        assertEquals(expected, a+b);
+    }
+}
Index: src/main/java/net/engineeringdigest/journalApp/config/SpringSecurity.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/engineeringdigest/journalApp/config/SpringSecurity.java b/src/main/java/net/engineeringdigest/journalApp/config/SpringSecurity.java
new file mode 100644
--- /dev/null	(date 1763196801566)
+++ b/src/main/java/net/engineeringdigest/journalApp/config/SpringSecurity.java	(date 1763196801566)
@@ -0,0 +1,42 @@
+package net.engineeringdigest.journalApp.config;
+
+import net.engineeringdigest.journalApp.services.UserDetailsServiceImpl;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
+@EnableWebSecurity
+public class SpringSecurity extends WebSecurityConfigurerAdapter {
+
+    @Autowired
+    private UserDetailsServiceImpl userDetailsService;
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.authorizeHttpRequests()
+                .antMatchers("/journal/**","/users/**").authenticated()
+                .antMatchers("/admin/**").hasRole("ADMIN")
+
+                .anyRequest().permitAll()
+                .and()
+                .httpBasic();
+         http.csrf().disable();
+    }
+
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth.userDetailsService(userDetailsService).passwordEncoder(passwordEncoder());
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder(){
+        return new BCryptPasswordEncoder();
+    }
+}
